src/main/java/org/opensearch/ad/settings/AbstractSetting.java

```diff
@@ -0,0 +1,97 @@
+ * GitHub history for details.
+ */
+
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package org.opensearch.ad.settings;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.settings.Setting;
+import org.opensearch.common.settings.Settings;
+
+/**
+ * A container serving dynamic numeric setting.  The caller does not have to call
+ * ClusterSettings.addSettingsUpdateConsumer and can access the most-up-to-date
+ * value using the singleton instance. This is convenient for a setting that's
+ * accessed by various places or it is not possible to install ClusterSettings.addSettingsUpdateConsumer
+ * as the enclosing instances are not singleton (i.e. deleted after use).
+ *
+ */
+public abstract class AbstractSetting {
+    private static Logger logger = LogManager.getLogger(AbstractSetting.class);
+
+    private ClusterService clusterService;
+    /** Latest setting value for each registered key. Thread-safe is required. */
+    private final Map<String, Object> latestSettings = new ConcurrentHashMap<>();
+
+    private final Map<String, Setting<?>> settings;
+
+    protected AbstractSetting(Map<String, Setting<?>> settings) {
+        this.settings = settings;
+    }
+
+    private void setSettingsUpdateConsumers() {
+        for (Setting<?> setting : settings.values()) {
+            clusterService.getClusterSettings().addSettingsUpdateConsumer(setting, newVal -> {
+                logger.info("[AD] The value of setting [{}] changed to [{}]", setting.getKey(), newVal);
+                latestSettings.put(setting.getKey(), newVal);
+            });
+        }
+    }
+
+    public void init(ClusterService clusterService) {
+        this.clusterService = clusterService;
+        setSettingsUpdateConsumers();
+    }
+
+    /**
+     * Get setting value by key. Return default value if not configured explicitly.
+     *
+     * @param key   setting key.
+     * @param <T> Setting type
+     * @return T     setting value or default
+     */
+    @SuppressWarnings("unchecked")
+    public <T> T getSettingValue(String key) {
+        return (T) latestSettings.getOrDefault(key, getSetting(key).getDefault(Settings.EMPTY));
```

Comment:
Can we put all setting's initial value into `latestSettings` map when `init` ? So we can just get setting values from `latestSettings`.

Reply:
If I understand you correctly, I have already done that.  init calls setSettingsUpdateConsumers that does the thing you asked.

Reply:
changed according to your suggestion. Please check the new commit.

Reply:
turns out all integration tests fail due to the above change.  I think it is because our rest service has no access to ClusterService and thus we didn't call AbstractSetting,init in our rest test case set up.

This is the changed code:
https://gist.github.com/kaituo/13ed77f2d7e2c92bc66877cc55d2b373

Then I tried to move latestSetting initialization to the constructor and comment out settings.  But there are compile errors.  Notice setting is of type Map<String, Setting<?>>, while latestSettings is of type Map<String, Object>. 

https://gist.github.com/kaituo/2b15a8279dd8534add5e6c69ffbcac15

Please advise how to proceed.

Reply:
Let's keep as is. Seems not easy to change.